File: my-solutions/0208-implement-trie-prefix-tree/solution.cpp
class TrieNode {
public:
    TrieNode* children[26];
    bool isWord;

    TrieNode() {
        memset(children, 0, sizeof(children));
        isWord = false;
    }
};

class Trie {
private:
    TrieNode* root;
    
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* node = root;
        for (char c : word) {
            if (!node->children[c - 'a']) {
                node->children[c - 'a'] = new TrieNode();
            }
            node = node->children[c - 'a'];
        }
        node->isWord = true;
    }
    
    bool search(string word) {
        TrieNode* node = searchPrefix(word);
        return node != nullptr && node->isWord;
    }
    
    bool startsWith(string prefix) {
        return searchPrefix(prefix) != nullptr;
    }
    
private:
    TrieNode* searchPrefix(string word) {
        TrieNode* node = root;
        for (char c : word) {
            if (!node->children[c - 'a']) return nullptr;
            node = node->children[c - 'a'];
        }
        return node;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: my-solutions/0208-implement-trie-prefix-tree/solution.py
class Trie:
    def __init__(self):
        self.root = TrieNode()
        

    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            if not node.containsKey(c):
                node.put(c, TrieNode())
            node = node.get(c)
        node.setEnd()
        

    def search(self, word: str) -> bool:
        node = self._searchPrefix(word)
        return node is not None and node.isEnd()
        

    def startsWith(self, prefix: str) -> bool:
        node = self._searchPrefix(prefix)
        return node is not None
    

    def _searchPrefix(self, prefix: str):
        node = self.root
        for c in prefix:
            if node.containsKey(c):
                node = node.get(c)
            else:
                return None
        return node
        


class TrieNode:
    def __init__(self):
        self.links = [None] * 26
        self.end = False
    

    def containsKey(self, c: str) -> bool:
        return self.links[ord(c) - ord('a')] is not None
    

    def get(self, c: str):
        return self.links[ord(c) - ord('a')]
    

    def put(self, c: str, node) -> None:
        self.links[ord(c) - ord('a')] = node
    

    def setEnd(self) -> None:
        self.end = True
    

    def isEnd(self) -> bool:
        return self.end



# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
